File: Asignaturas/Tercer Año/PDOO/Practica/Proyecto_Irrgarten/Proyecto_Irrgarten/SUBIR_A_PRADO_2_PARCIAL_PDOO/VARIOS_DISEÑOS/PROYECTO_SUPER_PLAYER/SolucionParcialPracticas2/P5-java/src/irrgarten/GameState.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

/**
 * La clase {@code GameState} representa el estado actual del juego, incluyendo
 * la información del laberinto, los jugadores, los monstruos, el jugador actual, 
 * y el registro de eventos.
 * 
 * <p>Esta clase es útil para almacenar y acceder al estado general del juego
 * durante el desarrollo de una partida.</p>
 * 
 * <p>Incluye detalles como:</p>
 * <ul>
 *   <li>Estado del laberinto y su disposición</li>
 *   <li>Jugadores y monstruos en el juego</li>
 *   <li>El jugador que tiene el turno actual</li>
 *   <li>Indicación de si existe un ganador</li>
 *   <li>Un registro de eventos que documenta el desarrollo de la partida</li>
 * </ul>
 * 
 * @author ismael
 */
public class GameState {
    
    /**
     * Estado actual del laberinto en forma de cadena.
     */
    static private String labyrinth;
    
    /**
     * Información sobre los jugadores en el juego en forma de cadena.
     */
    static private String players;
    
    /**
     * Información sobre los monstruos en el juego en forma de cadena.
     */
    static private String monster;
    
    /**
     * Índice del jugador que tiene el turno actual.
     */
    static private int currentPlayer;
    
    /**
     * Indica si existe un ganador en el juego.
     */
    static private boolean winner;
    
    /**
     * Registro de eventos en el juego en forma de cadena.
     */
    static private String log;
    
    /**
     * Crea una nueva instancia de {@code GameState} con el estado especificado del juego.
     *
     * @param labyrinth una representación en cadena del laberinto
     * @param players una representación en cadena de los jugadores
     * @param monster una representación en cadena del monstruo
     * @param currentPlayer el índice del jugador actual
     * @param winner {@code true} si hay un ganador; {@code false} en caso contrario
     * @param log registro de eventos en forma de cadena
     */
    public GameState(String labyrinth, String players, String monster, int currentPlayer, boolean winner, String log) {
        this.labyrinth = labyrinth;
        this.players = players;
        this.monster = monster;
        this.currentPlayer = currentPlayer;
        this.winner = winner;
        this.log = log;
    }
    
    /**
     * Obtiene la representación actual del laberinto.
     *
     * @return el estado del laberinto en forma de cadena
     */
    public String getlabyrinth() {
        return labyrinth;
    }
    
    /**
     * Obtiene la representación de los jugadores en el juego.
     *
     * @return los jugadores en forma de cadena
     */
    public String getplayers() {
        return players; 
    }
    
    /**
     * Obtiene la representación del monstruo en el juego.
     *
     * @return el monstruo en forma de cadena
     */
    public String getmonster() {
        return monster; 
    }
    
    /**
     * Obtiene el índice del jugador actual.
     *
     * @return el índice del jugador que tiene el turno
     */
    public int getcurrentPlayer() {
        return currentPlayer; 
    }
    
    /**
     * Indica si existe un ganador en el juego.
     *
     * @return {@code true} si hay un ganador; {@code false} en caso contrario
     */
    public boolean getwinner() {
        return winner; 
    }
    
    /**
     * Obtiene el registro de eventos del juego.
     *
     * @return el registro de eventos en forma de cadena
     */
    public String getlog() {
        return log; 
    }

    /**
     * Devuelve el estado de GameState en formato string
     * @return el estado de GameState en formato string
     */
    @Override
    public String toString() {
        return "\nlabyrinth=" + labyrinth +
            "\n \nplayers=" + players +
            "\n \nmonster=" + monster +
            "\n \ncurrentPlayer=" + currentPlayer +
            "\n \nwinner=" + winner +
            "\n \nlog=" + log +
            "\n";
        }
}

File: Asignaturas/Tercer Año/PDOO/Examenes/SolucionParcialPracticas2/P5-Ruby/lib/irrgarten/game_state.rb
#encoding:utf-8
module Irrgarten
    # Representa el estado actual del juego, incluyendo el laberinto, jugadores, monstruos,
    # el jugador actual, el ganador, y el registro de eventos (log).
    class GameState
        attr_accessor :labyrinth, :players, :monsters, :currentPlayer, :winner, :log

        # Inicializa un nuevo estado del juego.
        #
        # @param labyrinth [String] Representación del laberinto.
        # @param players [Array<Player>] Lista de jugadores en el juego.
        # @param monsters [Array<Monster>] Lista de monstruos en el juego.
        # @param currentPlayer [Player] Jugador que está en turno actualmente.
        # @param winner [Boolean] Estado de victoria, `true` si hay un ganador.
        # @param log [String] Registro de eventos ocurridos en el juego.
        def initialize(labyrinth, players, monsters, currentPlayer, winner, log)
            @labyrinth = labyrinth
            @players = players
            @monsters = monsters
            @currentPlayer = currentPlayer
            @winner = winner
            @log = log
        end

        # Retorna el laberinto actual.
        #
        # @return [String] Laberinto en su estado actual.
        def getlabyrinth
            @labyrinth
        end

        # Retorna la lista de jugadores.
        #
        # @return [Array<Player>] Lista de jugadores en el juego.
        def getplayers
            @players
        end

        # Retorna la lista de monstruos.
        #
        # @return [Array<Monster>] Lista de monstruos en el juego.
        def getmonsters
            @monsters
        end

        # Retorna el jugador actual en turno.
        #
        # @return [Player] Jugador en turno actual.
        def getcurrentPlayer
            @currentPlayer
        end

        # Indica si hay un ganador en el juego.
        #
        # @return [Boolean] `true` si hay un ganador, `false` si aún no.
        def getwinner
            @winner
        end

        # Retorna el log de eventos en el juego.
        #
        # @return [String] Log de eventos del juego.
        def getlog
            @log
        end

        def to_s
            "\nlabyrinth=#{labyrinth}" +
            "\n \nplayers=#{players}" +
            "\n \nmonsters=#{monsters}" +
            "\n \ncurrent_player=#{currentPlayer}" +
            "\n \nwinner=#{winner}" +
            "\n \nlog=#{log}" +
            "\n"
        end
            
    end
end
